spring-integration-zmq/src/main/java/org/springframework/integration/zmq/inbound/ZmqMessageDrivenChannelAdapter.java

```diff
@@ -0,0 +1,326 @@
+/*
+ * Copyright 2002-2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.zmq.inbound;
+
+import java.nio.charset.Charset;
+import java.util.Date;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledFuture;
+
+import org.zeromq.ZMQ;
+import org.zeromq.ZMQException;
+
+import org.springframework.context.ApplicationEventPublisher;
+import org.springframework.context.ApplicationEventPublisherAware;
+import org.springframework.integration.zmq.core.ConsumerStopAction;
+import org.springframework.integration.zmq.core.DefaultZmqClientFactory;
+import org.springframework.integration.zmq.core.ZmqClientFactory;
+import org.springframework.integration.zmq.event.ZmqConnectionFailedEvent;
+import org.springframework.integration.zmq.event.ZmqSubscribedEvent;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessagingException;
+import org.springframework.util.Assert;
+
+/**
+ * Jeromq Implementation.
+ *
+ * @author Subhobrata Dey
+ *
+ * @since 5.1
+ *
+ */
+public class ZmqMessageDrivenChannelAdapter extends AbstractZmqMessageDrivenChannelAdapter
+				implements ApplicationEventPublisherAware, Runnable {
+
+	private static final int DEFAULT_RECOVERY_INTERVAL = 10000;
+
+	private final ZmqClientFactory clientFactory;
+
+	private volatile ZMQ.Context context;
+
+	private volatile ZMQ.Socket client;
+
+	private volatile ZMQ.Poller poller;
+
+	private volatile ScheduledFuture<?> reconnectFuture;
+
+	private volatile boolean connected;
+
+	private volatile int recoveryInterval = DEFAULT_RECOVERY_INTERVAL;
+
+	private volatile boolean cleanSession;
+
+	private volatile ConsumerStopAction consumerStopAction;
+
+	private ApplicationEventPublisher applicationEventPublisher;
+
+	/**
+	 * Use this constructor for a single url (although it may be overridden
+	 * if the server URI is provided by the {@link DefaultZmqClientFactory}.
+	 * @param url the URL.
+	 * @param clientId The client id.
+	 * @param clientFactory The client factory.
+	 */
+	public ZmqMessageDrivenChannelAdapter(String url, String clientId,
+										ZmqClientFactory clientFactory) {
```

Comment:
I wonder if we really need that `url` as separate argument.

Maybe we could simplify things if we keep only a `clientFactory` abstraction?